src/tallies/filter_mesh.cpp

```diff
@@ -0,0 +1,132 @@
+#include "openmc/tallies/filter_mesh.h"
+
+#include <sstream>
+
+#include "openmc/capi.h"
+#include "openmc/constants.h"
+#include "openmc/error.h"
+#include "openmc/mesh.h"
+#include "openmc/xml_interface.h"
+
+namespace openmc {
+
+void
+MeshFilter::from_xml(pugi::xml_node node)
+{
+  auto bins_ = get_node_array<int32_t>(node, "bins");
+  if (bins_.size() != 1) {
+    fatal_error("Only one mesh can be specified per " + type()
+                + " mesh filter.");
+  }
+
+  auto id = bins_[0];
+  auto search = model::mesh_map.find(id);
+  if (search != model::mesh_map.end()) {
+    set_mesh(search->second);
+  } else{
+    std::stringstream err_msg;
+    err_msg << "Could not find mesh " << id << " specified on tally filter.";
+    fatal_error(err_msg);
+  }
+}
+
+void
+MeshFilter::get_all_bins(const Particle* p, int estimator, FilterMatch& match)
+const
+{
+  if (estimator != ESTIMATOR_TRACKLENGTH) {
+    auto bin = model::meshes[mesh_]->get_bin(p->r());
```

Comment:
I'll call it out again here, but generally not much of this code does parameter checks within functions, so for example `p` here could be `nullptr`. Since this function will crash if `p == nullptr`, it means that `p` is not optional, and so a reference may be a better choice than a pointer in order to enforce that and to document to developers that this is the case.